src/tga/encoder.rs

```diff
@@ -0,0 +1,125 @@
+    /// an error will be returned.
+    pub fn encode(
+        self,
+        buf: &[u8],
+        width: u32,
+        height: u32,
+        color_type: ColorType,
+    ) -> ImageResult<()> {
+        match color_type {
+            ColorType::Rgba8
+            | ColorType::Bgra8
+            | ColorType::Rgb8
+            | ColorType::Bgr8
+            | ColorType::La8
+            | ColorType::L8 => self
+                .encode_impl(buf, width, height, color_type)
+                .map_err(|err| {
+                    ImageError::Encoding(EncodingError::new(ImageFormat::Tga.into(), err))
+                }),
+            _ => Err(ImageError::Unsupported(
+                UnsupportedError::from_format_and_kind(
+                    ImageFormat::Tga.into(),
+                    UnsupportedErrorKind::Color(color_type.into()),
+                ),
+            )),
+        }
+    }
+
+    fn encode_impl(
+        mut self,
+        buf: &[u8],
+        width: u32,
+        height: u32,
+        color_type: ColorType,
+    ) -> io::Result<()> {
+        // Get pixel encoding data.
+        let (num_alpha_bits, other_channel_bits, image_type) = match color_type {
+            ColorType::Rgba8 | ColorType::Bgra8 => (8, 24, ImageType::RawTrueColor),
+            ColorType::Rgb8 | ColorType::Bgr8 => (0, 24, ImageType::RawTrueColor),
+            ColorType::La8 => (8, 8, ImageType::RawGrayScale),
+            ColorType::L8 => (0, 8, ImageType::RawGrayScale),
+            _ => unreachable!(),
+        };
+
+        // Write TGA header.
+        self.writer.write_u8(0)?; // No ID string.
+        self.writer.write_u8(0)?; // No color map.
+        self.writer.write_u8(image_type as u8)?; // Image type.
+        self.writer.write_u16::<LittleEndian>(0)?; // No color map.
+        self.writer.write_u16::<LittleEndian>(0)?; // No color map.
+        self.writer.write_u8(0)?; // No color map.
+        self.writer.write_u16::<LittleEndian>(0)?; // X-origin.
+        self.writer.write_u16::<LittleEndian>(0)?; // Y-origin.
+        self.write_dimension(width)?; // Width.
+        self.write_dimension(height)?; // Height.
+        self.writer.write_u8(num_alpha_bits + other_channel_bits)?; // Bits per pixel.
+        self.writer
+            .write_u8((num_alpha_bits & ALPHA_BIT_MASK) | SCREEN_ORIGIN_BIT_MASK)?; // Upper left origin.
+
+        // Write out Bgr(a)8 or L(a)8 image data.
+        let mut image = Vec::from(buf);
+
+        match color_type {
+            ColorType::Rgb8 | ColorType::Rgba8 => {
+                for chunk in image.chunks_mut(color_type.bytes_per_pixel() as usize) {
+                    chunk.swap(0, 2);
+                }
+            }
+            _ => {}
+        }
+
+        self.writer.write_all(&image)?;
+        Ok(())
+    }
+
+    fn write_dimension(&mut self, value: u32) -> io::Result<()> {
+        if value < 1 || value > MAX_IMAGE_DIMENSION {
```

Comment:
This could be checked before writing any data at all to avoid writing partial data. The violation should produce an `ImageError::EncodingError`, so checking it early would mean the writing can be left at a `io::Result` as well.

Reply:
I'm a little unclear on this one. I believe you're saying you want me to make something like a "check_dimension(value: u32) -> ImageResult" function and move the calls up to encode(), and you want me to refactor it to return something like this:
ImageError::Encoding(EncodingError::new(
            ImageFormat::Tga.into(),
            err,
        ));

So what kind of value will I be putting in "err"?

Reply:
That is what I intended to say, yes. First you would introduce a new private error enum with one variant for this kind of error. You could split it into two variants, one for width and height each, or handle both in the same variant. It would contain slightly more details to do two variants.

```rust
enum EncoderError {
    InvalidDimension(u32),
}
```

Then implement `fmt::Display` and `std::error::Error` for it. The displayed message would be what is currently being formatted. This allows you to pass an error of that type to `EncodingError::new`. The generic parameter and trait bound can be a little confusing at first (And for some reason the [`From` impl](https://doc.rust-lang.org/stable/src/std/error.rs.html#163) that enables this does currently not appear in the documentation of `std`. That's certainly not discoverable.).

Reply:
I did a little digging in the other encoders and found this pattern in a couple of them:
```rust
            u16::try_from(image.width()).map_err(|_| {
                ImageError::Parameter(ParameterError::from_kind(
                    ParameterErrorKind::DimensionMismatch,
                ))
            })?,
```
Are you sure I shouldn't be using something like this?

Reply:
It's more of a guideline, that's true. I imagine it's different in some places because previously that error was previously a top-level variant of `ImageError` and reworking it was a significant amount of work. The intended use case is:
* `EncodingError`: The format doesn't support some parameters.
* `UnsupportedError`: We don't support some parameters, but the format would.
* `ParameterError`: Everything not covered by the above. For example, encoding a bitmap only color values 0 and 255 must be provided.
Anyways, that's the intention. If you see any that you find differ from this explanation you can open an issue or add a TODO :)

Reply:
Okay, let me push my recent round of updates so you can take a look at that while I make those changes.

Reply:
Sorry, totally forgot about this part. I believe I was able to implement this the way you wanted, but I'll let you be the judge. Though it broke the ```test_image_too_large()``` function and I'm not familiar enough with Rust errors to fix it. Any useful insights about what to do there, or should I just leave it for now?

Reply:
Implemention looks indeed good. Let me get that test for you :)